NetKet/Sampler/abstract_sampler.hpp

```diff
@@ -50,8 +51,35 @@ class AbstractSampler {
     this->Reset(true);
   }
 
+  virtual void SetMachineFunc(
```

Comment:
Don't forget to check for empty state here! Also, you should probably use move constructor instead of copying. I.e.
```cpp
if (!machine_func) {
    throw SomeError{};
}
machine_func_ = std::move(machine_func);
```

P.S. Maybe even `machine_func_.swap(machine_func)` instead of using the assignment operator, because `operator=` is not `noexcept`, but `swap` is. And `machine_func` is destroyed anyway so it doesn't matter. Here, I'm assuming you got rid of the wrapper class as suggested in a comment below.

Reply:
~And `SetMachineFunc` shouldn't be virtual.~

Ah, `ExactSampler` overrides `SetMachineFunc`... For now, let's leave it `virtual` then, but `ExactSampler` should then also check for empty state. In the future, code duplication can be avoided by using an extra level of indirection, but it can be discussed in another PR since we're not claiming ABI stability.